File: services/eval/src/metrics/mrr.ts
/**
 * Reciprocal Rank for a single query.
 * The aggregator computes the mean across queries (MRR).
 */
export function meanReciprocalRank(retrievedIds: string[], relevantIds: string[]): number {
  if (relevantIds.length === 0) return 0;
  const relevantSet = new Set(relevantIds);

  for (let i = 0; i < retrievedIds.length; i++) {
    if (relevantSet.has(retrievedIds[i])) {
      return 1 / (i + 1);
    }
  }
  return 0;
}

File: services/eval/src/rag_eval/metrics/mrr.py
from collections.abc import Sequence


def reciprocal_rank(retrieved_ids: Sequence[str], relevant_ids: Sequence[str]) -> float:
    """Reciprocal rank of the first relevant hit for a single query.

    The aggregator computes the mean across queries (MRR).
    """
    if len(relevant_ids) == 0:
        return 0.0
    relevant_set = set(relevant_ids)

    for position, chunk_id in enumerate(retrieved_ids):
        if chunk_id in relevant_set:
            return 1 / (position + 1)
    return 0.0
